File: LeetCodePractice/src/graph/strongConnectComponent/CriticalConnectionsInaNetwork.java
package graph.strongConnectComponent;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/*There are n servers numbered from 0 to n-1 connected by undirected server-to-server connections forming a network where connections[i] = [a, b] represents a connection between servers a and b. Any server can reach any other server directly or indirectly through the network.
        A critical connection is a connection that, if removed, will make some server unable to reach some other server.
        Return all critical connections in the network in any order.

        Example 1:
        Input: n = 4, connections = [[0,1],[1,2],[2,0],[1,3]]
        Output: [[1,3]]
        Explanation: [[3,1]] is also accepted.

        Constraints:
        1 <= n <= 10^5
        n-1 <= connections.length <= 10^5
        connections[i][0] != connections[i][1]
        There are no repeated connections.

analysis:
Basically, Tarjan's Algorithm uses dfs to travel through the graph to find if current vertex u, can travel back to u or previous vertex
timeMap records the earliest vertex's time u can reach

TC O(N)
*/
public class CriticalConnectionsInaNetwork {
    public List<List<Integer>> criticalConnections(int n, List<List<Integer>> connections) {
        List<List<Integer>> res = new ArrayList<>();
        Map<Integer, Integer> timeMap = new HashMap<>();
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (List<Integer> connection : connections) {
            int start = connection.get(0);
            int end = connection.get(1);
            graph.computeIfAbsent(start, x -> new ArrayList<>()).add(end);
            graph.computeIfAbsent(end, x -> new ArrayList<>()).add(start);
        }
        dfs(graph, timeMap, res,0, -1, 0);
        return res;
    }

    private void dfs(Map<Integer, List<Integer>> graph, Map<Integer, Integer> timeMap, List<List<Integer>> res,
                     int cur, int pre, int time) {
        timeMap.put(cur, time);
        for (int nb : graph.get(cur)) {
            if (nb != pre) {
                if (!timeMap.containsKey(nb)) {
                    dfs(graph, timeMap, res, nb, cur, time + 1);
                }
                timeMap.put(cur, Math.min(timeMap.get(cur), timeMap.get(nb)));
                if (time < timeMap.get(nb)) {
                    res.add(Arrays.asList(cur, nb));
                }
            }
        }
    }
}

File: LeetCodePython/CriticalConnectionsInaNetwork.py
"""
There are n servers numbered from 0 to n - 1 connected by undirected server-to-server connections forming a network where connections[i] = [ai, bi] represents a connection between servers ai and bi. Any server can reach other servers directly or indirectly through the network.

A critical connection is a connection that, if removed, will make some servers unable to reach some other server.

Return all critical connections in the network in any order.

Example 1:
Input: n = 4, connections = [[0,1],[1,2],[2,0],[1,3]]
Output: [[1,3]]
Explanation: [[3,1]] is also accepted.

Example 2:
Input: n = 2, connections = [[0,1]]
Output: [[0,1]]

Constraints:
2 <= n <= 105
n - 1 <= connections.length <= 105
0 <= ai, bi <= n - 1
ai != bi
There are no repeated connections.

hint:
1 Use Tarjan's algorithm.

"""
import collections
from typing import List


class CriticalConnectionsInaNetwork:
    def criticalConnections(self, n: int, connections: List[List[int]]) -> List[List[int]]:
        res, time, graph = [], {}, collections.defaultdict(list)
        for s, e in connections:
            graph[s].append(e)
            graph[e].append(s)

        def dfs(cur, pre, t):
            time[cur] = t
            for nb in graph[cur]:
                if nb != pre:
                    if nb not in time:
                        dfs(nb, cur, t + 1)
                    time[cur] = min(time[cur], time[nb])
                    if t < time[nb]:
                        res.append([cur, nb])

        dfs(0, -1, 0)
        return res
